src/imageio/format/jxl.c

```diff
@@ -0,0 +1,740 @@
+        break;
+      case DT_COLORSPACE_LIN_REC709:
+        color_encoding.white_point = JXL_WHITE_POINT_D65;
+        primaries = &Rec709_Primaries;
+        color_encoding.transfer_function = JXL_TRANSFER_FUNCTION_LINEAR;
+        break;
+      case DT_COLORSPACE_ADOBERGB:
+        color_encoding.white_point = JXL_WHITE_POINT_D65;
+        primaries = &Adobe_Primaries;
+        color_encoding.transfer_function = JXL_TRANSFER_FUNCTION_GAMMA;
+        color_encoding.gamma = 2.19921875;
+        break;
+      case DT_COLORSPACE_PQ_P3:
+        color_encoding.white_point = JXL_WHITE_POINT_D65;
+        color_encoding.primaries = JXL_PRIMARIES_P3;
+        color_encoding.transfer_function = JXL_TRANSFER_FUNCTION_PQ;
+        break;
+      case DT_COLORSPACE_HLG_P3:
+        color_encoding.white_point = JXL_WHITE_POINT_D65;
+        color_encoding.primaries = JXL_PRIMARIES_P3;
+        color_encoding.transfer_function = JXL_TRANSFER_FUNCTION_HLG;
+        break;
+      case DT_COLORSPACE_PROPHOTO_RGB:
+        color_encoding.white_point = JXL_WHITE_POINT_D65;
+        primaries = &ProPhoto_Primaries;
+        color_encoding.transfer_function = JXL_TRANSFER_FUNCTION_LINEAR;
+        break;
+      default:
+        goto write_icc;
+    }
+
+    color_encoding.color_space = JXL_COLOR_SPACE_RGB;
+    if(primaries != NULL)
+    {
+      color_encoding.primaries = JXL_PRIMARIES_CUSTOM;
+      color_encoding.primaries_red_xy[0] = primaries->Red.x;
+      color_encoding.primaries_red_xy[1] = primaries->Red.y;
+      color_encoding.primaries_green_xy[0] = primaries->Green.x;
+      color_encoding.primaries_green_xy[1] = primaries->Green.y;
+      color_encoding.primaries_blue_xy[0] = primaries->Blue.x;
+      color_encoding.primaries_blue_xy[1] = primaries->Blue.y;
+    }
+    // TODO: safe to simply put pipe->icc_intent here?
+    color_encoding.rendering_intent = JXL_RENDERING_INTENT_PERCEPTUAL;
+    JxlEncoderSetColorEncoding(encoder, &color_encoding);
+    goto color_encoding_end;
+  }
+
+write_icc:
+  fprintf(stdout, "JXL: Could not generate color encoding data, falling back to ICC binary\n");
+
+  cmsUInt32Number size = 0;
+  unsigned char *icc_buf = NULL;
+  if(cmsSaveProfileToMem(out_profile, NULL, &size))
+  {
+    icc_buf = malloc(size);
+    if(icc_buf && cmsSaveProfileToMem(out_profile, icc_buf, &size))
+    {
+      JXL_ASSERT(JxlEncoderSetICCProfile(encoder, icc_buf, size));
+      free(icc_buf);
+      goto color_encoding_end;
+    }
+  }
+
+  if(icc_buf) free(icc_buf);
+  fprintf(stderr, "JXL: Error writing ICC data\n");
+  goto end;
+
+color_encoding_end:
+
+  JxlBasicInfo basic_info;
+  JxlEncoderInitBasicInfo(&basic_info);
+  basic_info.have_container = JXL_FALSE;
+  basic_info.xsize = width;
+  basic_info.ysize = height;
+  basic_info.bits_per_sample = 32;
+  basic_info.exponent_bits_per_sample = 8;
+  basic_info.intensity_target = 255.f;
+  basic_info.min_nits = 0.f;
+  basic_info.relative_to_max_display = JXL_FALSE;
```

Comment:
These I have no experience with:

- I thought `1.f` is the reference?
- How does `relative_to_max_display` [work out for the PQ transfer curve](https://github.com/darktable-org/darktable/issues/3880) case?

Reply:
This whole HDR business is so messy... Most photography/videography is still based on mid-gray anchoring, which makes 1.0 = "diffuse white" (for SDR this is the same as "max white"), but then permits values much greater than 1.0 for highlights. Unfortunately a lot of SW (and ICC) don't work this way, and I believe they assume 1.0 is max white (like you seem to do here as well). So not 100% sure what to do here, because from https://github.com/darktable-org/darktable/issues/3880 it sounds like darktable hasn't converged on how to handle this internally yet...

Side note: one is wasting a whole lot of dynamic range if limiting values to 0-1.0 only (still leaving plentiful 126 stops in the float case luckily), which is undoubtedly problematic if quantizing to float16 with that restriction (14 stops instead of available 30, so worse than 16-bit integer actually).